src/plugins/enzymes/src/FindEnzymesAlgorithm.h

```diff
@@ -51,29 +51,54 @@ class FindEnzymesAlgorithm {
+            } else if (enzyme->cutDirect > enzymeSequence.size()) {
+                trailingNsNumber = enzyme->cutDirect - enzymeSequence.size();
+            }
+        }
+        if (enzyme->cutComplement != ENZYME_CUT_UNKNOWN) {
+            if (enzyme->cutComplement < 0) {
+                trailingNsNumber = qMax(trailingNsNumber, qAbs(enzyme->cutComplement));
+            } else if (enzyme->cutComplement > enzymeSequence.size()) {
+                leadingNsNumber = qMax(leadingNsNumber, enzyme->cutComplement - enzymeSequence.size());
+            }
+        }
+        if (leadingNsNumber != 0) {
+            enzymeSequence.insert(0, QString(leadingNsNumber, EnzymeData::UNDEFINED_BASE));
+        }
+        if (trailingNsNumber != 0) {
+            enzymeSequence.append(QString(trailingNsNumber, EnzymeData::UNDEFINED_BASE));
+        }
+
         // look for results in direct strand
-        run(sequence, region, enzyme, enzyme->seq.constData(), U2Strand::Direct, resultListener, stateInfo, resultPosShift);
+        run(sequence, region, enzyme, enzymeSequence, U2Strand::Direct, resultListener, stateInfo, resultPosShift);
         CHECK_OP(stateInfo, );
 
         // if enzyme is not symmetric - look in complementary strand too
         DNATranslation* tt = AppContext::getDNATranslationRegistry()->lookupComplementTranslation(enzyme->alphabet);
         if (tt == nullptr) {
             return;
         }
-        QByteArray revCompl = enzyme->seq;
+        QByteArray revCompl = enzymeSequence;
         tt->translate(revCompl.data(), revCompl.size());
         TextUtils::reverse(revCompl.data(), revCompl.size());
-        if (revCompl == enzyme->seq) {
+        if (revCompl == enzymeSequence) {
             return;
         }
-        run(sequence, region, enzyme, revCompl.constData(), U2Strand::Complementary, resultListener, stateInfo, resultPosShift);
+        run(sequence, region, enzyme, revCompl, U2Strand::Complementary, resultListener, stateInfo, resultPosShift);
     }
 
-    void run(const DNASequence& sequence, const U2Region& region, const SEnzymeData& enzyme, const char* pattern, U2Strand stand, FindEnzymesAlgListener* resultListener, TaskStateInfo& ti, int resultPosShift = 0) {
+    void run(const DNASequence& sequence, const U2Region& region, const SEnzymeData& enzyme, const QByteArray& byteArrayPattern, U2Strand stand, FindEnzymesAlgListener* resultListener, TaskStateInfo& ti, int resultPosShift = 0) {
+        const char* pattern = byteArrayPattern.constData();
         CompareFN fn(sequence.alphabet, enzyme->alphabet);
         const char* seq = sequence.constData();
         char unknownChar = sequence.alphabet->getDefaultSymbol();
-        int plen = enzyme->seq.length();
+        int plen = byteArrayPattern.length();
         for (int pos = region.startPos, endPos = region.endPos() - plen + 1; pos < endPos; pos++) {
             bool match = matchSite(seq + pos, pattern, plen, unknownChar, fn);
```

Comment:
We have some problems with enzyme searching on circular molecules (see [UGENE-7908](https://ugene.dev/tracker/browse/UGENE-7908), reproducible on v. 47), so it would be better to fix it, and then figure out this scenario for circular molecules.

Reply:
It works for circular, I just tested it and everything is OK. I will add an XML test.

Reply:
I see, there is a special handling of circular sequences below